[DO NOT MERGE]

jamesr66a opened the pull request.

Commit:
init

Commit:
init

Commit:
cleanup

Commit:
file

zheng-xq committed:
A skeleton implmentation for the expression, IR and visitor dispatchers. (#33)

To run the test: cmake . && make cpptest && ./expr_test

jjsjann123 committed:
hacked CUDA add fusion

mruberry committed:
Delete ir.h

zheng-xq committed:
Refactor the RefHandle class. (#34)

Add convenience operator for Expr.

jjsjann123 committed:
patched the failed build from missing torch/csrc/jit/fuser/cpu/ir.h

mruberry committed:
Merge pull request #1 from jjsjann123/pytorch_fusion

hacked CUDA add fusion

zheng-xq committed:
clang-format change (#35)

zheng-xq committed:
Adding Var, Let and eval_context support. (#36)

bertmaher committed:
Add LLVM JIT class for online codegen

bertmaher committed:
Refactor llvm codegen

bertmaher committed:
fix caps of LlvmJit

bertmaher committed:
Generate code for integer arithmetic

bertmaher committed:
Test all arithmetic ops with LLVM

bertmaher committed:
Fix rtti

bertmaher committed:
Compat with llvm 7 and 8

zheng-xq committed:
Add support for tensor expressions. (#38)

Add Casting support so mixed dtypes are supported.
Add basic dtype and logging support. This should be merged with PyTorch during integration.

zheng-xq committed:
clang-format fix (#39)

zheng-xq committed:
Extend dtypes to support vector types (#40)

bertmaher committed:
Support LLVM 9 too

bertmaher committed:
Disambigate dependent type name with template keyword

bertmaher committed:
Remove empty scalar.h

zheng-xq committed:
Add basic support for statements. (#41)

Add support for For, Ramp, Block, Load, Store and Broadcast.
Add support for Buffer.

zheng-xq committed:
Adding Stmt evaluation support. (#42)

bertmaher committed:
Use third_party/googletest from pytorch

bertmaher committed:
Remove nnc/tests/googletest submodule

bertmaher committed:
Move nnc tld to torch/csrc/jit/compiler

bertmaher committed:
Add a README (probably temporary) for jit/compiler

bertmaher committed:
Move from namespace nnc to torch::jit::compiler

bertmaher committed:
Refactor JIT class to isolate no-rtti pieces

Commit:
Fix typo in README.md

Commit:
Use absolute imports and pragma once

zheng-xq committed:
Adding comparison operator to Var. (#43)

bertmaher committed:
Use absolute includes in new llvm_jit.h

Commit:
Build non-LLVM compiler stuff with libtorch

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 5621993:
* **2/2** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=562199354320a5ae919899f595b05e8f4b7d14a6).

### :detective: 2 new failures recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4011421) pytorch_linux_xenial_py3_6_gcc5_4_build (1/2)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=77127970) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=77127970))
<details>
<summary>
<code>Automatic merge failed; fix conflicts and then commit the result.</code>
</summary>

```
CONFLICT (add/add): Merge conflict in .circleci/scripts/binary_ios_upload.sh 
Auto-merging .circleci/scripts/binary_ios_upload.sh 
CONFLICT (add/add): Merge conflict in .circleci/scripts/binary_ios_build.sh 
Auto-merging .circleci/scripts/binary_ios_build.sh 
CONFLICT (add/add): Merge conflict in .circleci/generate_config_yml.py 
Auto-merging .circleci/generate_config_yml.py 
CONFLICT (add/add): Merge conflict in .circleci/docker/common/install_travis_python.sh 
Auto-merging .circleci/docker/common/install_travis_python.sh 
CONFLICT (add/add): Merge conflict in .circleci/config.yml 
Auto-merging .circleci/config.yml 
Automatic merge failed; fix conflicts and then commit the result. 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4011423) pytorch_xla_linux_xenial_py3_6_clang7_build (2/2)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=77129828) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=77129828))
<details>
<summary>
<code>Dec 19 01:23:38 RuntimeError: Failed to apply patch: /var/lib/jenkins/workspace/xla/torch_patches/X10-serialization.diff</code>
</summary>

```
Dec 19 01:23:38 + python /var/lib/jenkins/workspace/xla/scripts/cond_patch.py /var/lib/jenkins/workspace/xla/scripts/../torch_patches /var/lib/jenkins/workspace/xla/scripts/../.. 
Dec 19 01:23:38 1 out of 1 hunk FAILED 
Dec 19 01:23:38 Applying patch file: /var/lib/jenkins/workspace/xla/torch_patches/X10-clip_grad.diff 
Dec 19 01:23:38 Applying patch file: /var/lib/jenkins/workspace/xla/torch_patches/X10-device_test.diff 
Dec 19 01:23:38 Applying patch file: /var/lib/jenkins/workspace/xla/torch_patches/X10-serialization.diff 
Dec 19 01:23:38 Traceback (most recent call last): 
Dec 19 01:23:38   File "/var/lib/jenkins/workspace/xla/scripts/cond_patch.py", line 67, in <module> 
Dec 19 01:23:38     patch_repo(args) 
Dec 19 01:23:38   File "/var/lib/jenkins/workspace/xla/scripts/cond_patch.py", line 49, in patch_repo 
Dec 19 01:23:38     raise RuntimeError('Failed to apply patch: {}'.format(ppath)) 
Dec 19 01:23:38 RuntimeError: Failed to apply patch: /var/lib/jenkins/workspace/xla/torch_patches/X10-serialization.diff 
Dec 19 01:23:38 + cleanup 
Dec 19 01:23:38 + retcode=1 
Dec 19 01:23:38 + set +x 
Dec 19 01:23:38 =================== sccache compilation log =================== 
Dec 19 01:23:38 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Dec 19 01:23:38 Compile requests                 1 
Dec 19 01:23:38 Compile requests executed        0 
Dec 19 01:23:38 Cache hits                       0 
Dec 19 01:23:38 Cache misses                     0 
Dec 19 01:23:38 Cache timeouts                   0 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

bertmaher committed:
LLVM code generation for simple loops

Commit:
Minimal asmjit codegen from the tensor IR

Commit:
Merge pull request #3 from jamesr66a/pytorch_fusion

Minimal asmjit codegen from the tensor IR

Commit:
fix pessimizing moves

Commit:
Merge pull request #4 from jamesr66a/pytorch_fusion

fix pessimizing moves

jjsjann123 committed:
[Fuser]Fuser registration

1. allow dynamic registration of FusionBackend
2. supports cpu-only/gpu-only builds.

Commit:
Merge pull request #5 from jjsjann123/pytorch_fusion_separate_build_pr

[Fuser]Fuser registration

Commit:
IR printer

Commit:
Merge pull request #6 from jamesr66a/pytorch_fusion

IR printer

Commit:
fix printer bug

Commit:
Merge pull request #7 from jamesr66a/pytorch_fusion

fix printer bug

zheng-xq committed:
Add data structure for schedule support and Split.

csarofeen committed:
Add printer to build system.

csarofeen committed:
First pass at copying the IR over.

Commit:
cpp test extensibility point

Commit:
Removes (most) old code, updates README

Commit:
comments out test_fuser

csarofeen committed:
Typo in build_variables

zheng-xq committed:
Merge pull request #8 from zheng-xq/tec_01_merge

Add data structure for schedule support and Split.

zheng-xq committed:
clang-format using the new template

csarofeen committed:
Take a pass at the type system. Add simple print test.

csarofeen committed:
Merge branch 'pytorch_fusion' of https://github.com/mruberry/pytorch into tmp

zheng-xq committed:
Add IRMutator and basic support to substitude Var in Expr and Stmts.

Commit:
Build fixes

Commit:
Buildable basics 1/X

Commit:
Merge branch 'pytorch_fusion' of https://github.com/mruberry/pytorch into pytorch_fusion

csarofeen committed:
Take first pass at container class for fusions, make things more const and strict.

zheng-xq committed:
Change the default count of RefCounted as zero.
Merge Expr(node) and Expr::make(node).

Commit:
Buildable basics 2/X

Commit:
small cleanup

Commit:
Start of (pretty) printing

Commit:
Symbolic floats, Add outputs, more (pretty) printing

csarofeen committed:
Fuser cleanup.

csarofeen committed:
A bit more cleanup, some commenting.

Commit:
removes old files

csarofeen committed:
Start moving to Exprs only having Vals as inputs/outputs

csarofeen committed:
Fix Fusion inFusion check.

csarofeen committed:
Another quick safety check in fusion.

csarofeen committed:
Add implicit arith, Int, and basic type promotion.

csarofeen committed:
First pass at lifetime model.

csarofeen committed:
Bit of manager cleanup.

zheng-xq committed:
Add basic lowering to the tensor expression trees.

csarofeen committed:
Improve printing of a fusion.

csarofeen committed:
Clean up debug spew.

csarofeen committed:
Test addition, organization and cleanup.

zheng-xq committed:
fix the schedule_test

zheng-xq committed:
fixed lowering

bertmaher committed:
bugfixes

zheng-xq committed:
refcount fixing self-assignment

zheng-xq committed:
Make LOG(FATAL) nonreturn
Enable Werror

zheng-xq committed:
Adding statement conversion for SplitWithTail

zheng-xq committed:
Add a reference tests for Split

bertmaher committed:
clang-format

zheng-xq committed:
A functinoal reference chck for schedule tests.

zheng-xq committed:
clang-format

Commit:
Add support for Float immediates.

bertmaher committed:
Get absolute path for ASMJIT_DIR (#24)

bertmaher committed:
Silence deprecation warnings from LLVM

bertmaher committed:
Include legacy PassManager for debug printing

bertmaher committed:
Set code model to medium to avoid indirect jumps in generated asm

bertmaher committed:
Fix argument type of input float buffers

Commit:
Add support for Casts in LLVM codegen.

Commit:
Enable export of compile_commands.json.

zheng-xq committed:
Add a complete tensor+lower+llvm test

zheng-xq committed:
Enable the failing test

bertmaher committed:
Floating point arithmetic

bertmaher committed:
Test fp32 mul using compute expr

bertmaher committed:
Broadcast add test using compute expr

Commit:
Update to LLVM 9

Commit:
Implementation of Broadcast for LLVM.

Commit:
Cleanup use of ConstantInt API.

zheng-xq committed:
Add Buffer operator() overload, and some other minor features

zheng-xq committed:
fix accidental experimental changes

zheng-xq committed:
Change the Compute interface to bring the dim sizes and names together

zheng-xq committed:
clang-format

zheng-xq committed:
refactor Buffer into its own files

zheng-xq committed:
Add a PaddedBuffer test util

Commit:
Add support for vector casts in LLVM CodeGen

Commit:
Implement masked loads and stores.

Commit:
Implement vector masked loads and stores.
